src/lambda/wizards/samDeployWizard.ts

```diff
@@ -357,23 +391,66 @@ export class DefaultSamDeployWizardContext implements SamDeployWizardContext {
                     'AWS.samcli.deploy.s3Bucket.prompt',
                     'Enter the AWS S3 bucket to which your code should be deployed'
                 ),
+                value: initialValue || '',
+                matchOnDetail: true,
                 ignoreFocusOut: true,
-                prompt: localize(
-                    'AWS.samcli.deploy.s3Bucket.region',
-                    'S3 bucket must be in selected region: {0}',
-                    selectedRegion
+            },
+            // preloaded "busy screen" message
+            // using a const in case we want to make this clickable as a way to abort the lookup.
+            items: [
+                {
+                    label: loadingBuckets
+                }
+            ],
+        })
+
+        // throw up a busy screen while the S3 buckets are loading
+        picker.makeQuickPickBusy(quickPick)
+        quickPick.show()
+
+        try {
+            const listBucketsResponse = await s3Client.listBuckets()
+
+            // sort through buckets and add to cache if you had to run a listBuckets
+            if (listBucketsResponse && listBucketsResponse.Buckets) {
+                // listBuckets lists all buckets tied to an account regardless of region
+                // we need to filter the list ourselves since we do care about the region
+                const bucketsForRegion = await getBucketsForRegion(
+                    listBucketsResponse.Buckets,
+                    selectedRegion,
+                    s3Client
                 )
+                quickPick.items = createBucketQuickPickItems(bucketsForRegion, initialValue)
             }
-        })
 
-        // Pre-populate the value if it was already set
-        if (initialValue) {
-            inputBox.value = initialValue
+            // no buckets found for region
+            if (quickPick.items.length === 0) {
+                quickPick.items = [
+                    {
+                        label: s3NoBucketsLabel,
+                        description: s3GoBack
+                    }
+                ]
+            }
+        } catch (err) {
+            // user doesn't have access to region for some reason
+            const error = err as Error
+            const logger = getLogger()
+            logger.warn(error)
+            quickPick.items = [
+                {
+                    label: s3ErrorLabel,
+                    description: s3GoBack,
+                    detail: error.message
+                }
+            ]
         }
 
-        return await input.promptUser({
-            inputBox: inputBox,
-            onValidateInput: validateS3Bucket,
+        // loading is done, restore to a selectable state
+        picker.makeQuickPickNotBusy(quickPick)
+
+        const pickedItem =  await picker.promptUser<vscode.QuickPickItem>({
```

Comment:
If you press `Esc` while the buckets are loading, the picker goes away. This causes two problems:
* the wizard didn't go "back" one step
* the buckets are still loading. Once loading completes, the picker pops up. Surprise!

`quickPick.onDidHide` might be useful in accounting for this.

Reply:
Do we want to define a standard behavior for this case? I'm assuming we want `QuickPick.hide()` to be handled as a `return undefined`?

Reply:
✔️ although I'm unsure of the implementation. Looking for feedback on whether this is the correct path before I add tests to cover this.